### PR TITLE
Add the possibility to co-install multiple kubectl versions

### PR DESCRIPTION
**Pull request by mssola**
# Describe your changes

Adds some documentation on the possibility to co-install multiple `kubectl` versions, plus some minor improvements (e.g. consistency on the management pattern being installed).

# Related Issues / Projects

SUSE/avant-garde#1878

### CODE REVIEW

File: adoc/deployment-bootstrap.adoc [hunk 1]
```diff
@@ -798,3 +798,32 @@ kube-system   kured-zv696                             1/1       Running   0
...
+
+This can be useful, for example, when managing multiple Kubernetes clusters from
+the same {productname} management machine. That is, following {productname} 4.5,
+{productname} will be able to manage other (supported) Kubernetes versions than
+only 1.17.
+
+For example, in a system like this:
+
+[source,bash]
+----
```
Comment (mssola) [inline adoc/deployment-bootstrap.adoc:817]: @r0ckarong I don't like that this code block does not have horizontal scrolling. Is there any way in AsciiDoc to force this?

File: adoc/deployment-bootstrap.adoc [hunk 2]
```diff
@@ -798,3 +798,32 @@ kube-system   kured-zv696                             1/1       Running   0
...
+Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.4", GitCommit:"8d8aa39598534325ad77120c120a22b3a990b5ea", GitTreeState:"clean", BuildDate:"2020-03-17T12:00:00Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"linux/amd64"}
+----
+
+You can install another version and use it right away with `update-alternatives`:
+
+[source,bash]
+----
+# zypper install kubernetes-1.18-client
+# update-alternatives --set kubectl /usr/bin/kubectl-1.18
+----
```
Comment (mssola) [inline adoc/deployment-bootstrap.adoc:829]: I'm still stuck on checking that this actually works on a real machine, so take it with a grain of salt (I won't be around tomorrow nor next week so I wanted to submit this PR anyways)

File: adoc/deployment-bootstrap.adoc [hunk 3]
```diff
@@ -798,3 +798,32 @@ kube-system   kured-zv696                             1/1       Running   0
 kube-system   oidc-dex-55fc689dc-b9bxw                1/1       Running   0          2m
 kube-system   oidc-gangway-7b7fbbdbdf-ll6l8           1/1       Running   0          2m
 ----
+
+==== Co-installing multiple versions of kubectl
+
+It's possible to install multiple versions of `kubectl` on the same machine. In
+a {productname} management machine, you will be able to co-install multiple
+versions of `kubectl`, and later mark which is the version to be used with
+`update-alternatives`.
```
Comment (jodavis-suse) [inline adoc/deployment-bootstrap.adoc:807]: You may consider mentioning that by default the kubectl with the highest version number will be selected.

File: adoc/deployment-bootstrap.adoc (hunk 2 above)
Comment (jodavis-suse) [reply adoc/deployment-bootstrap.adoc:829]: Two notes here:
First, we are not delivering a 1.17 client that will be co-installable with 1.18.  The 1.17 packaging was written without update-alternatives so isn't compatible with that functionality, so the 1.17 "kubernetes-client" package will be removed when "kuberenetes-1.18-client" is installed.
Second, by default the highest version number will be selected by default by update-alternatives when installed.  So in the scenario described, 1.18 would already be selected and you would not need to run the "update-alternatives --set kubectl /usr/bin/kubectl-1.18" command to use it.  If, after installing 1.19 then the user wanted to go back to 1.18, then they would use that command.

Comment (evrardjp) [reply adoc/deployment-bootstrap.adoc:829]: I agree with @jodavis-suse . Next to this, I would prefer if we don't encourage people to install the client on the master/worker nodes, as this is handled by skuba. We don't want users to mess up the expected versions there. In order to clarify that this is for the management machine, I would mention `You can install another version *on a management node* and use it right away with `update-alternatives` on L823

File: adoc/deployment-bootstrap.adoc (hunk 1 above)
Comment (r0ckarong) [reply adoc/deployment-bootstrap.adoc:817]: The final output will have horizontal scrolling. This depends on the settings of the generator processing the input files. What are you using to preview this?

**Comment by jordimassaguerpla** [comment]
@mssola @jodavis-suse : Please correct me if I am wrong. This feature is not part of GMC3 but in our Devel project, correct?

GMC3: https://build.suse.de/package/show/SUSE:SLE-15-SP2:Update:Products:CaaSP:4.5/kubernetes-1.18
Devel: https://build.suse.de/project/show/Devel:CaaSP:4.5

**Comment by jodavis-suse** [comment]
> @mssola @jodavis-suse : Please correct me if I am wrong. This feature is not part of GMC3 but in our Devel project, correct?
> 
> GMC3: https://build.suse.de/package/show/SUSE:SLE-15-SP2:Update:Products:CaaSP:4.5/kubernetes-1.18
> Devel: https://build.suse.de/project/show/Devel:CaaSP:4.5

This change (to kubernetes-1.18-client) very much SHOULD be part of GMC3 and in the release.  I thought there was already a plan in place to include it, as #1813 has been closed.

**Review by evrardjp — APPROVED** [review]

**Comment by evrardjp** [comment]
s/SHOULD/MUST/

File: adoc/deployment-bootstrap.adoc (hunk 3 above)
Comment (mjura) [reply adoc/deployment-bootstrap.adoc]: Added

File: adoc/deployment-bootstrap.adoc (hunk 2 above)
Comment (mjura) [reply adoc/deployment-bootstrap.adoc]: Fixed

**Review by chentex — APPROVED** [review]

**Comment by jordimassaguerpla** [comment]
Hi. The "update-alternatives" change in k8s package is not in GMC.

See:

https://github.com/SUSE/avant-garde/issues/1813#issuecomment-660978091 

> ... "multiple kubectl versions" is not a 5.0.0 release criteria. If it's easy to get it, I'd welcome its inclusion. But I wouldn't wait for it.

https://github.com/SUSE/avant-garde/issues/1813#issuecomment-668606841
> removing the gmc2 requirement in favor of #1868 

In GMC3 we only include P1 issues. This was never considered as this nor planned for GMC3.

See also https://github.com/SUSE/avant-garde/issues/1914#issuecomment-677468813

**Comment by jodavis-suse** [comment]
> Hi. The "update-alternatives" change in k8s package is not in GMC.
> 
> See:
> 
> [SUSE/avant-garde#1813 (comment)](https://github.com/SUSE/avant-garde/issues/1813#issuecomment-660978091)
> 
> > ... "multiple kubectl versions" is not a 5.0.0 release criteria. If it's easy to get it, I'd welcome its inclusion. But I wouldn't wait for it.
> 
> [SUSE/avant-garde#1813 (comment)](https://github.com/SUSE/avant-garde/issues/1813#issuecomment-668606841)
> 
> > removing the gmc2 requirement in favor of #1868
> 
> In GMC3 we only include P1 issues. This was never considered as this nor planned for GMC3.
> 
> See also [SUSE/avant-garde#1914 (comment)](https://github.com/SUSE/avant-garde/issues/1914#issuecomment-677468813)

I'm going to reread all the comments, but please understand this is very frustrating from a developer standpoint. The change has been waiting for multiple process changes (going through the new CI branching process then the whole 5.0->4.5 rename) and keeps getting bumped out rather than smoothly included.

File: adoc/deployment-bootstrap.adoc [hunk 4]
```diff
@@ -798,3 +798,39 @@ kube-system   kured-zv696                             1/1       Running   0
...
+# kubectl version
+Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.6", GitCommit:"dff82dc0de47299ab66c83c626e08b245ab19037", GitTreeState:"clean", BuildDate:"2020-08-04T12:00:00Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
+Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.6", GitCommit:"dff82dc0de47299ab66c83c626e08b245ab19037", GitTreeState:"clean", BuildDate:"2020-08-04T12:00:00Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
+----
+
+You can install another version and use other version previously installed with `update-alternatives`:
+
+[source,bash]
+----
+# zypper install kubernetes-1.19-client
```
Comment (jodavis-suse) [inline adoc/deployment-bootstrap.adoc:827]: I just noticed on line 746 this document has the command line 'sudo zypper install kubernetes-client' which is not consistent with this line.

**Comment by r0ckarong** [comment]
@jordimassaguerpla @davidcassany Which version is this going the be included with?

**Comment by r0ckarong** [comment]
@kkaempf @davidcassany Is this still relevant?

**Comment by kkaempf** [comment]
> @kkaempf @davidcassany Is this still relevant?

Not immediately, but will come up again when we move `kubectl` to SLE. Public cloud team will need a different `kubectl` version that CaaSP for example.

**Comment by davidcassany** [comment]
@r0ckarong I did a submissionf for k8sclt on SLE-15-SP2 of `kubernetes1.18-client-1.18.10`. This is a multi installable package in the sense that it makes use of update-alternatives to permit the user to switch from one version to another. This will happen from 1.18 and on (on minor level version, aka 1.18, 1.19, 1.20, etc.).

Note small detail on packaging name `kubernetes1.18-client` is expected to be the new one, up to now in v4.5 series it was `kubernetes-1.18-client` (note the dash difference).

@r0ckarong I am not sure I am answering your concerns, let me know if not.

**Comment by mssola** [comment]
Closing since this PR might no longer be relevant.

**Pull request closed without merging**